composer/datasets/in_context_learning_evaluation.py

```diff
@@ -70,6 +81,155 @@ def _get_fewshot_sample_idxs(dataset_size, num_fewshot, sample_idx):
+            if dist.get_local_rank() == 0:
+                get_file(dataset_uri, destination_path, overwrite=True)
+        dataset = load_dataset('json', data_files=destination_path, split='train', streaming=False)
+        self.samples = list(
+            dataset.map(lambda examples: {
+                'context': examples['context'],
+                'answer': examples['answer'],
+                'aliases': examples['aliases']
+            }))
+        self.tokenizer = tokenizer
+        self.max_seq_len = max_seq_len
+        self.pad_tok_id = pad_tok_id
+        self.max_answer_length = 0
+        self.encoded_dataset = self.prep_examples(num_fewshot, prompt_string, example_delimiter, continuation_delimiter,
+                                                  question_prelimiter)
+
+    def prep_examples(self, num_fewshot: int, prompt_string: str, example_delimiter: str, continuation_delimiter: str,
+                      question_prelimiter: str):
+        """Prepares a set of language modeling tasks into tokenized format with prompt and fewshot examples.
+
+        Each task consists of a context and a continuation as well as an optional prompt and optional list of
+        example context/continuation pairs which precede the test context/continuation pair.
+
+        Args:
+            num_fewshot (int): Number of examples context/continuation pairs to prepend to the test pair
+            prompt_string (str): The prompt to prepend to all inputs
+            example_delimiter (str): The delimiter used to separate each individual context/continuation pair
+            continuation_delimiter (str): The delimiter used to separate each context from its continuation
+            question_prelimiter (str): The text to prepend to each question
+
+        Returns:
+            dict: Contains the context, the continuation, and the preamble (prompt + fewshot examples)
+        """
+        max_answer_length = 0
+        examples = []
+        for sample_idx in tqdm(range(len(self.samples))):
+            encoded_example = {}
+
+            preamble = prompt_string
+
+            if num_fewshot > 0:
+                fewshot_idxs = _get_fewshot_sample_idxs(len(self.samples), num_fewshot, sample_idx)
+                for fewshot_idx in fewshot_idxs:
+                    ctxt, cont = self.samples[fewshot_idx]['context'], self.samples[fewshot_idx]['answer']
+                    ctxt = f'{question_prelimiter}{ctxt}'
+                    if len(preamble) > 0:
+                        ctxt = f'{example_delimiter}{ctxt}'
+                    preamble += f'{ctxt}{continuation_delimiter}{cont}'
+
+            ctxt = self.samples[sample_idx]['context']
+            ctxt = f'{question_prelimiter}{ctxt}'
+            if len(preamble) > 0:
+                ctxt = f'{example_delimiter}{ctxt}'
+
+            # rstrip the continuation delimiter, because the prompt ending in a space results in degenerate output
```

Comment:
Are you serious? that's absurd...

Reply:
🤷‍♂️ not sure if it's a thing that is consistent across models, or if I was doing something else wrong. I should probably try to double check this before merging.

Reply:
Yeah, that is definitely the reason, I'm just not completely sure if the prompt _not_ ending in a space is a thing that all models prefer.